src/plugins/usage_collection/server/usage_counters/usage_counters_service.ts

```diff
@@ -0,0 +1,178 @@
+  SavedObjectsServiceStart,
+} from 'src/core/server';
+import type { Logger } from 'src/core/server';
+
+import moment from 'moment';
+import { CounterMetric, UsageCounter } from './usage_counter';
+import {
+  registerUsageCountersSavedObjectType,
+  storeCounter,
+  serializeCounterKey,
+} from './saved_objects';
+
+export interface UsageCountersServiceDeps {
+  logger: Logger;
+  retryCount: number;
+  bufferDebounceMs: number;
+}
+
+export interface UsageCountersServiceSetup {
+  createUsageCounter: (type: string) => UsageCounter;
+  getUsageCounterByType: (type: string) => UsageCounter | undefined;
+}
+
+/* internal */
+export interface UsageCountersServiceSetupDeps {
+  savedObjects: SavedObjectsServiceSetup;
+}
+
+/* internal */
+export interface UsageCountersServiceStartDeps {
+  savedObjects: SavedObjectsServiceStart;
+}
+
+export class UsageCountersService {
+  private readonly stop$ = new Rx.Subject();
+  private readonly retryCount: number;
+  private readonly bufferDebounceMs: number;
+
+  private readonly counterSets = new Map<string, UsageCounter>();
+  private readonly source$ = new Rx.Subject<CounterMetric>();
+  private readonly counter$ = this.source$.pipe(rxOp.multicast(new Rx.Subject()), rxOp.refCount());
+  private readonly flushCache$ = new Rx.Subject<CounterMetric>();
+
+  private readonly stopCaching$ = new Rx.Subject();
+
+  private readonly logger: Logger;
+
+  constructor({ logger, retryCount, bufferDebounceMs }: UsageCountersServiceDeps) {
+    this.logger = logger;
+    this.retryCount = retryCount;
+    this.bufferDebounceMs = bufferDebounceMs;
+  }
+
+  public setup = (core: UsageCountersServiceSetupDeps): UsageCountersServiceSetup => {
+    const cache$ = new Rx.ReplaySubject<CounterMetric>();
+    const storingCache$ = new Rx.BehaviorSubject<boolean>(false);
+    // flush cache data from cache -> source
+    this.flushCache$
+      .pipe(
+        rxOp.exhaustMap(() => cache$),
+        rxOp.takeUntil(this.stop$)
+      )
+      .subscribe((data) => {
+        storingCache$.next(true);
+        this.source$.next(data);
+      });
+
+    // store data into cache when not paused
+    storingCache$
+      .pipe(
+        rxOp.distinctUntilChanged(),
+        rxOp.switchMap((isStoring) => (isStoring ? Rx.EMPTY : this.source$)),
+        rxOp.takeUntil(Rx.merge(this.stopCaching$, this.stop$))
+      )
+      .subscribe((data) => {
+        cache$.next(data);
+        storingCache$.next(false);
+      });
+
+    registerUsageCountersSavedObjectType(core.savedObjects);
+
+    return {
+      createUsageCounter: this.createUsageCounter,
+      getUsageCounterByType: this.getUsageCounterByType,
+    };
+  };
+
+  public start = ({ savedObjects }: UsageCountersServiceStartDeps): void => {
+    this.stopCaching$.next();
+    const internalRepository = savedObjects.createInternalRepository();
+    this.counter$
+      .pipe(
+        rxOp.buffer(this.source$.pipe(rxOp.debounceTime(this.bufferDebounceMs))),
```

Comment:
why not using `rxOp.bufferTime`?

However, I'm wondering how safe this is. Won't we lose data on restarts? What's the benefit from calling SO.incrementCounter in batches when we are actually looping through the counters anyway?

Wouldn't that create more spikes every 30s instead of spreading them?

Maybe I'm not reading the code correctly and I missed on something. What do you think?

Reply:
`bufferTime` behaves differently that what i'm indending here


> Won't we lose data on restarts?

We'd be losing 30 seconds of data (or whatever we set `bufferDebounceMs` to). I believe this is not something we should be worried about since its a small duration. If you think otherwise we can use `leading: true` to send it at the beginning of the interval instead of at the end of it.

>  What's the benefit from calling SO.incrementCounter in batches when we are actually looping through the counters anyway?

We need some buffer time in case the same counter gets called multiple times in a row; The reduce function will group them in 1 single SO call instead of multiple ones to the same key.

We dont have 1 batch operation for incrementing multiple counters at once. Maybe we can do this as an enhancement in the future to make this even more efficient.

> Wouldn't that create more spikes every 30s instead of spreading them?

I dont think it would make a huge difference. I've reduced the `30s` to `5s` following your suggestion in the thread below.